integration_test/hostmetrics_test.go

```diff
@@ -0,0 +1,304 @@
+
+	if len(result) < 2 {
+		t.Fatalf("Unreasonably few (<2) expectations found. expectations=%v", result)
+	}
+	return result
+}
+
+// loadObservedMetrics reads the contents of metrics.json produced by
+// otelopscol and selects one of the lines from it, corresponding to a single
+// batch of exported metrics.
+func loadObservedMetrics(t *testing.T, metricsJSONPath string) pmetric.Metrics {
+	data, err := os.ReadFile(metricsJSONPath)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	lines := strings.Split(string(data), "\n")
+	if len(lines) < 2 {
+		t.Fatalf("Only found %v lines in %q, need at least 2", len(lines), metricsJSONPath)
+	}
+
+	// Take the second batch of data exported by otelopscol. Picking the first
+	// batch can hit problems with certain metrics like system.cpu.utilization,
+	// which don't appear in the first batch.
+	secondBatch := []byte(lines[1])
+
+	t.Logf("Found %v bytes of data at %s, selecting %v bytes", len(data), metricsJSONPath, len(secondBatch))
+
+	metrics, err := pmetric.NewJSONUnmarshaler().UnmarshalMetrics(secondBatch)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return metrics
+}
+
+// expectMetricsMatch checks that all the data in `observedMetrics` matches
+// the expectations configured in `expectedMetrics`.
+// Note that an individual metric can appear many times in `observedMetrics`
+// under different values of its resource attributes. In particular, the
+// process.* metrics have resource attributes and so they will appear N times,
+// where N is the number of process resources that were detected.
+func expectMetricsMatch(t *testing.T, observedMetrics pmetric.Metrics, expectedMetrics map[string]ExpectedMetric) {
+	// Holds the set of metrics that were seen somewhere in observedMetrics.
+	seen := make(map[string]bool)
+
+	resourceMetrics := observedMetrics.ResourceMetrics()
+	for i := 0; i < resourceMetrics.Len(); i++ {
+		scopeMetrics := resourceMetrics.At(i).ScopeMetrics()
+
+		resourceAttributes := resourceMetrics.At(i).Resource().Attributes() // Used below.
+		for j := 0; j < scopeMetrics.Len(); j++ {
+			metrics := scopeMetrics.At(j).Metrics()
+
+			for k := 0; k < metrics.Len(); k++ {
+				metric := metrics.At(k)
+
+				name := metric.Name()
+
+				seen[name] = true
+
+				expected, ok := expectedMetrics[name]
+				if !ok {
+					// It's debatable whether a new metric being introduced should cause
+					// this test to fail. We decided to fail the test in this case because
+					// otherwise, there's no incentive to add coverage for new metrics as
+					// they appear. If it proves onerous to keep fixing this test, we can
+					// remove this check and come up with some other way to keep this test
+					// up to date.
+					t.Errorf("Unexpected metric %q observed.", name)
+					continue
+				}
+
+				if metric.Type().String() != expected.Type {
+					t.Errorf("For metric %q, Type()=%v, want %v",
+						name,
+						metric.Type().String(),
+						expected.Type,
+					)
+				}
```

Comment:
In general, yes, I think it's better to report all errors that we can. Otherwise you have no idea how close to passing your test is. You end up in a soul-sucking loop of fixing one thing and then another error pops up, and you don't know how many more hidden errors there will be.

Regarding t.Run, we could do that I suppose, but we are visiting the process metrics several times due to resource attributes, so it's not super clean. I introduced some errors into the config file and ran the test with and without t.Run, you tell me which you like better, personally I prefer the run without t.Run:

http://go/paste/5307401250013184
http://go/paste/5062361353289728

Reply:
I quite like the output with t.Run just cause I like how easy it is at a glance to see errors grouped together by which expected metric they happened on. The run without it looks far more readable than I envisioned though, so I'm fine with either.

Reply:
ok added t.Run.